src/client/Microsoft.Identity.Client/Credential/TokenRequestAssertionInfo.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Platforms.Features.KeyMaterial;
+
+namespace Microsoft.Identity.Client.Credential
+{
+    /// <summary>
+    /// Class to store crypto key information for a Managed Identity supported Azure resource.
+    /// For more details see https://aka.ms/msal-net-managed-identity
+    /// </summary>
+    internal class TokenRequestAssertionInfo
+    {
+        private readonly X509Certificate2 _bindingCertificate;
+        private readonly CertificateCache _certificateCache;
+        private readonly object _keyInfoLock = new(); // Lock object
+        private readonly ILoggerAdapter _logger;
+
+        public TokenRequestAssertionInfo(IKeyMaterialManager keyMaterialManager,
+            IServiceBundle serviceBundle)
+        {
+            _logger = serviceBundle.ApplicationLogger;
+
+            _certificateCache = CertificateCache.Instance();
+
+            _bindingCertificate = _certificateCache.GetOrAddCertificate(
+                () => CreateCertificateFromCryptoKeyInfo(keyMaterialManager));
+        }
+
+        public X509Certificate2 BindingCertificate => _bindingCertificate;
+
+        private X509Certificate2 CreateCertificateFromCryptoKeyInfo(IKeyMaterialManager keyMaterialManager)
+        {
+            lock (_keyInfoLock) // Lock to ensure thread safety
```

Comment:
Consider using the pattern below, which avoids locking in most cases.

`if (_bindingCertificate == null)
       lock
            if (_bindingCertificate == null)
                  _bindingCertificate  = create()`